proof_of_stake/src/lib.rs

```diff
@@ -0,0 +1,1814 @@
+            &mut validator_total_deltas,
+            &mut validator_voting_power,
+            &mut total_voting_power,
+            &mut validator_set,
+            current_epoch,
+        )?;
+
+        let total_bonds = bond.get_at_offset(
+            current_epoch,
+            DynEpochOffset::PipelineLen,
+            &params,
+        );
+        match total_bonds {
+            Some(total_bonds) if total_bonds.sum() != 0.into() => {
+                self.write_bond(&bond_id, bond);
+            }
+            _ => {
+                // If the bond is left empty, delete it
+                self.delete_bond(&bond_id)
+            }
+        }
+        self.write_unbond(&bond_id, unbond);
+        self.write_validator_total_deltas(validator, validator_total_deltas);
+        self.write_validator_voting_power(validator, validator_voting_power);
+        self.write_total_voting_power(total_voting_power);
+        self.write_validator_set(validator_set);
+
+        Ok(())
+    }
+
+    /// Withdraw unbonded tokens from a self-bond to a validator when `source`
+    /// is `None` or equal to the `validator` address, or withdraw unbonded
+    /// tokens delegated to the `validator` to the `source`.
+    fn withdraw_tokens(
+        &mut self,
+        source: Option<&Self::Address>,
+        validator: &Self::Address,
+        current_epoch: impl Into<Epoch>,
+    ) -> Result<Self::TokenAmount, WithdrawError<Self::Address>> {
+        let current_epoch = current_epoch.into();
+        let params = self.read_pos_params();
+        let source = source.unwrap_or(validator);
+        let bond_id = BondId {
+            source: source.clone(),
+            validator: validator.clone(),
+        };
+
+        let unbond = self.read_unbond(&bond_id);
+        let slashes = self.read_validator_slashes(&bond_id.validator);
+
+        let WithdrawData {
+            unbond,
+            withdrawn,
+            slashed,
+        } = withdraw_unbonds(
+            &params,
+            &bond_id,
+            unbond,
+            slashes,
+            current_epoch,
+        )?;
+
+        let total_unbonds = unbond.get_at_offset(
+            current_epoch,
+            DynEpochOffset::UnbondingLen,
+            &params,
+        );
+        match total_unbonds {
+            Some(total_unbonds) if total_unbonds.sum() != 0.into() => {
+                self.write_unbond(&bond_id, unbond);
+            }
+            _ => {
+                // If the unbond is left empty, delete it
+                self.delete_unbond(&bond_id)
+            }
+        }
+
+        // Transfer the tokens from PoS back to the source
+        self.transfer(
+            &Self::staking_token_address(),
+            withdrawn,
+            &Self::POS_ADDRESS,
+            source,
+        );
+
+        Ok(slashed)
+    }
+}
+
+/// PoS system base trait for system initialization on genesis block, updating
+/// the validator on a new epoch and applying slashes.
+pub trait PosBase {
```

Comment:
It seems like most of (if not all) of the methods and types from the above two traits are also here. If so, can we simplify by making `pub trait PosBase: PosReadOnly + PosActions`?

Reply:
I have considered this, but I there are some differences and we would end-up with unused functions in the `PosBase` that shouldn't even be used from base :( We started we a single trait and because of the different integration layers, ended up with all of these. It's not very pretty, so I'm thinking it might be to refactor this completely without the traits.